python unit-testing module - closes #678 (#679)
bob-crutchley/courseware: topics/python/modules/unit-testing/README.md

<!--TOC_END-->
## Overview

In this module we will cover the basics of unit testing with Python. We will talk about testing in general and then move on to using `pytest`.

## The Theory Of Testing

Testing (According to ANSI/IEEE 1059 standard) 'is a process of analysing a software item to detect the differences between existing and required conditions (i.e. defects) and to evaluate the features of the software item'.

Simply, testing is the activity of running code under various scenarios and circumstances to highlight if and where defects exist. As well as discover the applications characteristics, in comparison to business requirements and rules.

### What can go wrong?

Imagine we are copying a file from one location to another. This is a simple task that people do everyday. But lets consider what can go wrong with this task.

- File could not be found
- Destination unavailable
- File already in use
- Incorrect permissions
- Network fail during transmission
- File corrupted during transmission
- Disk space availability
- Blocked by firewall

So one simple task and all these possible errors, imagine all the possible errors for 200 lines of code.

### Why do we test?

> The cost of fixing defects when a product is in production (live) is exponentially more expensive then code defects identified in an earlier phase.

Reviewer: ```suggestion
The cost of fixing defects when a product is in production (live) is far greater than the cost of fixing defects identified in an earlier phase.
```

> Releasing buggy code, often, can have impact on customer faith and business reputation.

Reviewer: ```suggestion
Releasing buggy code can often have an impact on customer faith and business reputation.
```

Testing aids in providing quality assurance to a business.

### Types of testing

There are three different types of testing and many methods to carry out the different types.

![types](https://qa-courseware-images.s3.eu-west-2.amazonaws.com/python/unit-testing/000.png)

- `Functional` - If I click a button, does it perform the correct action?
- `Non-Functional` - If 10,000 people click the same button, is the response time acceptable? 
- `Maintenance` - Once changes have been made, do all previous tests still pass?

## Unit testing in Python

### Why use automated testing?

There are many reasons why we should `automate` tests, the main reason is `repeatability`. 

> We can run tests `whenever` a change is made and not have to completely rewrite the tests from scratch. And be confident that if the changes pass the tests, your software will work with `minimal defects`.

Reviewer: ```suggestion
We can run tests `whenever` a change is made and not have to completely rewrite the tests from scratch each time. We can also be confident that if the changes pass the tests, your software will work with `minimal defects`.
```

The easiest way to achieve this is through unit tests.
This is a scripted test in Python to verify unit of functionality.

Python accommodates for this through a number of tools.
We will be focusing on `pytest`. 

### Pytest introduction

> Pytest is a framework that allows for automated testing in Python. Installing it is a simple `pip` install in `Linux`.

Reviewer: pip installs work in Windows cmd, with the standard Python download comes the pip package manager - the command is the same

> Pytest is a framework that allows for automated testing in Python. Installing it is a simple `pip` install in `Linux`.

Author: ```suggestion
Pytest is a framework that allows for automated testing in Python. Installing it is a simple `pip` install.
```

```bash
pip install pytest
```

Within your code you will have to `import pytest` and then you can start using the tool.
This is a very simple example. What do you think this test is doing?

```python
import pytest

def func(num):
    return num * 2

def test_answer():
    assert func(6) == 10
```

<details>
<summary>Answer</summary>

This test is running the function `func` and passing the parameter `6` and checking if you get a return of `10`.
</details>

### Asserting in pytest

`Assert` statements allow us to establish whether a test has passed or failed by asserting certain states of the code.
For example, you can write the following, will this test pass?

```python
def f():
    return 3

def test_function():
    assert f() == 4
```

No, this test will not pass, as the `f()` function is going to return `3` not `4`.

### Writing tests

When you run the `pytest` command, it will look for filenames in for the for `test_[name].py` or `[name]_test.py`.

> In them files your tests must be in the form of functions and similarly to the filename the name of the function must be in the form `def [name]_test():` or `def test_[name]():`.

Reviewer: ```suggestion
In these files your tests must be in the form of functions and similarly to the filename, the name of the function must be in the form `def [name]_test():` or `def test_[name]():`.
```

We want to see how much of our application has been tested and to what extend it has been tested.
This is another simple `pip` install.

```bash
pip install pytest-cov
```
You then run the `pytest` command with some parameters.

```bash
pytest [testfile] --cov=[file to test]
```

## Tutorial

We are going to write a Python program and write the tests for the program.

Create 2 new python files within the same directory. Call one `count.py` and the other `test_count.py`.

In `count.py` enter the following.

```python
def count(sequence, item):
  sum = 0
  for n in sequence:
    if n == item:
      sum += 1
  return sum
```

We are going to write 2 tests for this function. We are gonna test if the function works for integers as well as strings.

Enter the following into `test_count.py`

```python
import pytest
from python_exercises import count

def test_count_zeros():
    assert count.count([0,0,0], 0) == 3

def test_count_string():
    assert count.count(["a","a","a"], "a") == 3
```

Now run the command:

```bash
pytest
```

You should be able to see the tests have passed.

## Exercises

> Now go to [Here](https://www.github.com/jordon-grindrod/python_exercises) and write tests for 3 of these Python files and use test coverage to achieve `100% coverage`.s

Reviewer: ```suggestion
Now go to [Here](https://www.github.com/jordan-grindrod/python_exercises) and write tests for 3 of these Python files and use test coverage to achieve `100% coverage`.s
```
